src/lib.rs

```diff
@@ -180,9 +180,18 @@ impl<F: PrimeField, PC: PolynomialCommitment<F>, D: Digest> Marlin<F, PC, D> {
 
         let prover_init_state = AHPForR1CS::prover_init(&index_pk.index, c, zk)?;
         let public_input = prover_init_state.public_input();
-        let mut fs_rng = PC::RandomOracle::from_seed(
-            &to_bytes![&Self::PROTOCOL_NAME, pc_pk.get_hash(), &index_pk.index_vk, &public_input].unwrap(),
-        );
+
+        let fs_rng_init_seed = {
+            let mut seed_builder = <PC::RandomOracle as FiatShamirRng>::Seed::new();
+            seed_builder.add_bytes(&Self::PROTOCOL_NAME).map_err(Error::from_pc_err)?;
+            seed_builder.add_bytes(&pc_pk.get_hash()).map_err(Error::from_pc_err)?;
+            // TODO: Shall we decompose this further when passing it to the seed builder ?
+            seed_builder.add_bytes(&index_pk.index_vk).map_err(Error::from_pc_err)?;
+            // TODO: Shall we decompose this further when passing it to the seed builder ?
+            seed_builder.add_bytes(&public_input).map_err(Error::from_pc_err)?;
```

Comment:
In regards of the ``TODO``: What would be the reason for decomposing it further?

Reply:
If we consider the public_inputs/vk as a single entity contributing to the seed, then the internal num_elements counter will just be incremented by one and we will add just one elem length, that is the total length of the public_inputs/vk; instead we can consider each input of the public inputs/each element of the vk singularly, so the internal num_elements counter will be incremented by public_inputs.len()/vk.num_elements() and public_inputs.len()/vk.num_elements() elements will be pushed to the internal element_length array.
We will end up with two different seeds, but which choice would be the most cryptographically sound ?